# Gesture Recognition
In this group project, you are going to build a 3D Conv model that will be able to predict the 5 gestures correctly. Please import the following libraries to get started.

In [2]:
from google.colab import drive
drive.mount('/content/drive/')


Mounted at /content/drive/


In [3]:
# temp code
# import os

# def count_files_and_folders(directory):
#     num_files = 0
#     num_folders = 0

#     for root, dirs, files in os.walk(directory):
#         num_folders += len(dirs)
#         num_files += len(files)

#     return num_files, num_folders
# # /content/drive/My Drive/AI/Upgrad/Melanoma_Assignment_CNN/CNN_assignment/Skin cancer ISIC The International Skin Imaging Collaboration/Train
# # Example usage
# directory = '/content/drive/My Drive/AI/Upgrad/Gesture_Recognition_CaseStudy/Project_data/Project_data/train'  # Replace with your directory path
# files, folders = count_files_and_folders(directory)
# print(f"Number of files: {files}")
# print(f"Number of folders: {folders}")

# directory = '/content/drive/My Drive/AI/Upgrad/Gesture_Recognition_CaseStudy/Project_data/Project_data/val'  # Replace with your directory path
# files, folders = count_files_and_folders(directory)
# print(f"Number of files: {files}")
# print(f"Number of folders: {folders}")

In [35]:
import numpy as np
import os
import matplotlib.pyplot as plt

# from scipy.misc import imread, imresize
from imageio import imread
from PIL import Image
# Resize using PIL.Image.resize
# image = Image.fromarray(image)  # Convert to PIL Image
# image = image.resize((100, 100))  # Resize to (100, 100)
# image = np.array(image)  # Convert back to NumPy array

import datetime
import os

from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Conv3D, MaxPooling3D , Flatten, Dropout, GlobalAveragePooling3D, BatchNormalization

from tensorflow.keras.applications import ResNet50
from tensorflow.keras.applications.resnet50 import preprocess_input

from tensorflow.keras.layers import TimeDistributed, GRU

from tensorflow.keras.applications import ResNet50,MobileNetV2
from tensorflow.keras.layers import Input, TimeDistributed, GRU, Dense,GlobalAveragePooling2D
from tensorflow.keras.models import Model

from tensorflow.keras import regularizers

In [5]:
#!pip install --upgrade tensorflow

We set the random seed so that the results don't vary drastically.

In [6]:
np.random.seed(30)
import random as rn
rn.seed(30)
from keras import backend as K
import tensorflow as tf

#tf.set_random_seed(30)
tf.random.set_seed(30)

In this block, you read the folder names for training and validation. You also set the `batch_size` here. Note that you set the batch size in such a way that you are able to use the GPU in full capacity. You keep increasing the batch size until the machine throws an error.

In [7]:
train_doc = np.random.permutation(open('/content/drive/My Drive/AI/Upgrad/Gesture_Recognition_CaseStudy/Project_data/Project_data/train.csv').readlines())
val_doc = np.random.permutation(open('/content/drive/My Drive/AI/Upgrad/Gesture_Recognition_CaseStudy/Project_data/Project_data/val.csv').readlines())

#experiment with the batch size
batch_size = 50


In [8]:
type(train_doc)

numpy.ndarray

## Generator
This is one of the most important part of the code. The overall structure of the generator has been given. In the generator, you are going to preprocess the images as you have images of 2 different dimensions as well as create a batch of video frames. You have to experiment with `img_idx`, `y`,`z` and normalization such that you get high accuracy.

In [9]:
#source_path = '/content/drive/My Drive/AI/Upgrad/Gesture_Recognition_CaseStudy/Project_data/Project_data'
#folder_list = train_doc,val_doc above
#batch_size = batch_size above
def generator(source_path, folder_list, batch_size):
    # print( 'Source path = ', source_path, '; batch size =', batch_size)
    img_idx = [i for i in range(0,30,2)]
    # img_idx = [i for i in range(0,30,1)]
    while True:
        # print('Here 1')
        # print(folder_list)

        t = np.random.permutation(folder_list) #temp comment

        # calculate the number of batches
        num_batches = int(len(folder_list)/batch_size)
        # print('num_batches',num_batches)

        for batch in range(num_batches): # we iterate over the number of batches
            # print('111')
            batch_data = np.zeros((batch_size,len(img_idx),120,160,3)) # x is the number of images you use for each video, (y,z) is the final size of the input images and 3 is the number of channels RGB
            batch_labels = np.zeros((batch_size,5)) # batch_labels is the one hot representation of the output
            for folder in range(batch_size): # iterate over the batch_size
                # print('222')
                imgs = os.listdir(source_path+'/'+ t[folder + (batch*batch_size)].split(';')[0]) # read all the images in the folder
                # print(t[folder])
                imgs = os.listdir(source_path+'/'+ t[folder + (batch*batch_size)].split(';')[0]) # read all the images in the folder
                for idx,item in enumerate(img_idx): #  Iterate iver the frames/images of a folder to read them in
                    # print('333')
                    image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)
                    #crop the images and resize them. Note that the images are of 2 different shape
                    #and the conv3D will throw error if the inputs in a batch have different shapes
                    image = Image.fromarray(image.astype(np.uint8))  # Convert to PIL Image
                    image = image.resize((160, 120))  # Resize to (100, 100)
                    image = np.array(image)  # Convert back to NumPy array

                    # print(image[:,:,0],image[:,:,0],image[:,:,2])
                    batch_data[folder,idx,:,:,0] = image[:,:,0]/255.0
                    batch_data[folder,idx,:,:,1] = image[:,:,1]/255.0
                    batch_data[folder,idx,:,:,2] = image[:,:,2]/255.0

                batch_labels[folder, int(t[folder + (batch*batch_size)].strip().split(';')[2])] = 1
            yield batch_data, batch_labels #you yield the batch_data and the batch_labels, remember what does yield do

        # write the code for the remaining data points which are left after full batches
        remaining_data_points = len(folder_list) - batch_size*num_batches
        batch_data = np.zeros((batch_size,len(img_idx),120,160,3)) # x is the number of images you use for each video, (y,z) is the final size of the input images and 3 is the number of channels RGB
        batch_labels = np.zeros((batch_size,5)) # batch_labels is the one hot representation of the output
        for folder in range(len(folder_list)-remaining_data_points,len(folder_list)): # iterate over the batch_size
            counter = 0
            # print('remaining: ',t[folder])
            imgs = os.listdir(source_path+'/'+ t[folder].split(';')[0]) # read all the images in the folder
            for idx,item in enumerate(img_idx): #  Iterate iver the frames/images of a folder to read them in
                image = imread(source_path+'/'+ t[folder].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)

                #crop the images and resize them. Note that the images are of 2 different shape
                #and the conv3D will throw error if the inputs in a batch have different shapes
                image = Image.fromarray(image.astype(np.uint8))  # Convert to PIL Image
                image = image.resize((160,120))  # Resize to (100, 100)
                image = np.array(image)  # Convert back to NumPy array

                # print(image[:,:,0]/255.0,image[:,:,0]/255.0,image[:,:,2]/255.0)
                batch_data[counter,idx,:,:,0] = image[:,:,0]/255.0
                batch_data[counter,idx,:,:,1] = image[:,:,1]/255.0
                batch_data[counter,idx,:,:,2] = image[:,:,2]/255.0

            batch_labels[counter, int(t[counter].strip().split(';')[2])] = 1
            counter += 1
        yield batch_data, batch_labels #you yield the batch_data and the batch_labels, remember what does yield do





In [9]:
# # temp code
# train_doc2 = open('/content/drive/My Drive/AI/Upgrad/Gesture_Recognition_CaseStudy/Project_data/temp/train.csv').readlines()
# val_doc2 = open('/content/drive/My Drive/AI/Upgrad/Gesture_Recognition_CaseStudy/Project_data/temp/val.csv').readlines()

In [10]:
# # temp code
# source_path = '/content/drive/My Drive/AI/Upgrad/Gesture_Recognition_CaseStudy/Project_data/temp/train'
# # source_path = '/content/drive/My Drive/AI/Upgrad/Gesture_Recognition_CaseStudy/Project_data/Project_data/train'
# folder_list = train_doc2
# batch_size = 4
# generator(source_path, folder_list, batch_size)

Note here that a video is represented above in the generator as (number of images, height, width, number of channels). Take this into consideration while creating the model architecture.

In [20]:
# curr_dt_time = datetime.datetime.now()
train_path = '/content/drive/My Drive/AI/Upgrad/Gesture_Recognition_CaseStudy/Project_data/Project_data/train'
val_path = '/content/drive/My Drive/AI/Upgrad/Gesture_Recognition_CaseStudy/Project_data/Project_data/val'
num_train_sequences = len(train_doc)
print('# training sequences =', num_train_sequences)
num_val_sequences = len(val_doc)
print('# validation sequences =', num_val_sequences)
num_epochs = 20
print ('# epochs =', num_epochs)

# training sequences = 663
# validation sequences = 100
# epochs = 20


## Model
Here you make the model using different functionalities that Keras provides. Remember to use `Conv3D` and `MaxPooling3D` and not `Conv2D` and `Maxpooling2D` for a 3D convolution model. You would want to use `TimeDistributed` while building a Conv2D + RNN model. Also remember that the last layer is the softmax. Design the network in such a way that the model is able to give good accuracy on the least number of parameters so that it can fit in the memory of the webcam.

In [21]:
from keras.models import Sequential, Model
from keras.layers import Dense, GRU, Flatten, TimeDistributed, Flatten, BatchNormalization, Activation,Conv3D, MaxPooling3D, GlobalAveragePooling3D, Dropout
# from keras.layers.convolutional import Conv3D, MaxPooling3D
from keras.callbacks import ModelCheckpoint, ReduceLROnPlateau
from keras import optimizers

#write your model here


In [22]:
img_idx = [i for i in range(0,30,2)]
# img_idx = [i for i in range(0,30,1)]
len(img_idx)

15

In [23]:
def fitmodel(model):
  curr_dt_time = datetime.datetime.now()
  optimiser = 'adam'
  model.compile(optimizer=optimiser, loss='categorical_crossentropy', metrics=['categorical_accuracy'])
  print (model.summary())

  train_generator = generator(train_path, train_doc, batch_size)
  val_generator = generator(val_path, val_doc, batch_size)

  model_name = 'model_init' + '_' + str(curr_dt_time).replace(' ','').replace(':','_') + '/'

  if not os.path.exists(model_name):
      os.mkdir(model_name)
      print(f"Model directory: {os.path.abspath(model_name)}")

  filepath = model_name + 'model-{epoch:05d}-{loss:.5f}-{categorical_accuracy:.5f}-{val_loss:.5f}-{val_categorical_accuracy:.5f}.keras'

  checkpoint = ModelCheckpoint(filepath, monitor='val_loss', verbose=1, save_best_only=False, save_weights_only=False, mode='auto', save_freq='epoch')

  LR = ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=3, verbose=1, mode='auto', min_lr=0.00001)
  callbacks_list = [checkpoint, LR]

  if (num_train_sequences%batch_size) == 0:
      steps_per_epoch = int(num_train_sequences/batch_size)
  else:
      steps_per_epoch = (num_train_sequences//batch_size) + 1

  if (num_val_sequences%batch_size) == 0:
      validation_steps = int(num_val_sequences/batch_size)
  else:
      validation_steps = (num_val_sequences//batch_size) + 1

  model.fit(train_generator, steps_per_epoch=steps_per_epoch, epochs=num_epochs, verbose=0,
                      callbacks=callbacks_list, validation_data=val_generator,
                      validation_steps=validation_steps, class_weight=None,  initial_epoch=0)


### Model Experiment1

In [ ]:
# model = Sequential()

# model.add(Conv3D(32, (3,3,3), padding='same', activation='relu', input_shape=(len(img_idx),120, 160, 3)))
# model.add(MaxPooling3D(pool_size=(2, 2, 2)))
# model.add(Dropout(0.25))

# model.add(Conv3D(64, (3,3,3), padding='same', activation='relu'))
# model.add(MaxPooling3D(pool_size=(2, 2, 2)))
# model.add(Dropout(0.25))

# model.add(Conv3D(128, (3,3,3), padding='same', activation='relu'))
# model.add(MaxPooling3D(pool_size=(2, 2, 2)))
# model.add(Dropout(0.25))

# model.add(Conv3D(256, (3,3,3), padding='same', activation='relu'))
# model.add(MaxPooling3D(pool_size=(2, 2, 2)))
# model.add(Dropout(0.25))

# # model.add(Flatten())
# model.add(GlobalAveragePooling3D())
# model.add(Dense(256, activation='relu'))
# model.add(Dropout(0.5))
# model.add(Dense(5, activation='softmax'))

Now that you have written the model, the next step is to `compile` the model. When you print the `summary` of the model, you'll see the total number of parameters you have to train.

In [ ]:
# optimiser = 'adam'
# model.compile(optimizer=optimiser, loss='categorical_crossentropy', metrics=['categorical_accuracy'])
# print (model.summary())

Let us create the `train_generator` and the `val_generator` which will be used in `.fit_generator`.

In [ ]:
# train_generator = generator(train_path, train_doc, batch_size)
# val_generator = generator(val_path, val_doc, batch_size)

In [ ]:
# model_name = 'model_init' + '_' + str(curr_dt_time).replace(' ','').replace(':','_') + '/'

# if not os.path.exists(model_name):
#     os.mkdir(model_name)

# filepath = model_name + 'model-{epoch:05d}-{loss:.5f}-{categorical_accuracy:.5f}-{val_loss:.5f}-{val_categorical_accuracy:.5f}.keras'

# checkpoint = ModelCheckpoint(filepath, monitor='val_loss', verbose=1, save_best_only=False, save_weights_only=False, mode='auto', save_freq='epoch')

# LR = ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=3, verbose=1, mode='auto', min_lr=0.00001)
# callbacks_list = [checkpoint, LR]

The `steps_per_epoch` and `validation_steps` are used by `fit_generator` to decide the number of next() calls it need to make.

In [ ]:
# if (num_train_sequences%batch_size) == 0:
#     steps_per_epoch = int(num_train_sequences/batch_size)
# else:
#     steps_per_epoch = (num_train_sequences//batch_size) + 1

# if (num_val_sequences%batch_size) == 0:
#     validation_steps = int(num_val_sequences/batch_size)
# else:
#     validation_steps = (num_val_sequences//batch_size) + 1

Let us now fit the model. This will start training the model and with the help of the checkpoints, you'll be able to save the model at the end of each epoch.

In [ ]:
# model.fit(train_generator, steps_per_epoch=steps_per_epoch, epochs=num_epochs, verbose=0,
#                     callbacks=callbacks_list, validation_data=val_generator,
#                     validation_steps=validation_steps, class_weight=None,  initial_epoch=0)

### Model too complex , running for too much time , simplifying

### Model Experiment2

In [ ]:
model = Sequential()

model.add(Conv3D(32, (3,3,3), padding='same', activation='relu', strides=(1, 2, 2), input_shape=(len(img_idx),120, 160, 3)))
model.add(MaxPooling3D(pool_size=(2, 2, 2)))
model.add(Dropout(0.25))

model.add(Conv3D(64, (3,3,3), padding='same', activation='relu', strides=(1, 2, 2)))
model.add(MaxPooling3D(pool_size=(2, 2, 2)))
model.add(Dropout(0.25))

# model.add(Conv3D(128, (3,3,3), padding='same', activation='relu'))
# model.add(MaxPooling3D(pool_size=(2, 2, 2)))
# model.add(Dropout(0.25))

# model.add(Conv3D(256, (3,3,3), padding='same', activation='relu'))
# model.add(MaxPooling3D(pool_size=(2, 2, 2)))
# model.add(Dropout(0.25))

# model.add(Flatten())
model.add(GlobalAveragePooling3D())
model.add(Dense(256, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(5, activation='softmax'))



In [ ]:
model.summary()

Model: "sequential_9"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv3d_15 (Conv3D)          (None, 30, 60, 80, 32)    2624      
                                                                 
 max_pooling3d_15 (MaxPooli  (None, 15, 30, 40, 32)    0         
 ng3D)                                                           
                                                                 
 dropout_22 (Dropout)        (None, 15, 30, 40, 32)    0         
                                                                 
 conv3d_16 (Conv3D)          (None, 15, 15, 20, 64)    55360     
                                                                 
 max_pooling3d_16 (MaxPooli  (None, 7, 7, 10, 64)      0         
 ng3D)                                                           
                                                                 
 dropout_23 (Dropout)        (None, 7, 7, 10, 64)     

In [ ]:
fitmodel(model)

Model: "sequential_9"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv3d_15 (Conv3D)          (None, 30, 60, 80, 32)    2624      
                                                                 
 max_pooling3d_15 (MaxPooli  (None, 15, 30, 40, 32)    0         
 ng3D)                                                           
                                                                 
 dropout_22 (Dropout)        (None, 15, 30, 40, 32)    0         
                                                                 
 conv3d_16 (Conv3D)          (None, 15, 15, 20, 64)    55360     
                                                                 
 max_pooling3d_16 (MaxPooli  (None, 7, 7, 10, 64)      0         
 ng3D)                                                           
                                                                 
 dropout_23 (Dropout)        (None, 7, 7, 10, 64)     

<ipython-input-31-c0c78d1fc080>:29: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)
<ipython-input-31-c0c78d1fc080>:54: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)



Epoch 1: saving model to model_init_2025-01-0305_13_55.125308/model-00001-1.50215-0.17714-1.59780-0.23000.keras

Epoch 2: saving model to model_init_2025-01-0305_13_55.125308/model-00002-1.49243-0.20143-1.59729-0.23000.keras

Epoch 3: saving model to model_init_2025-01-0305_13_55.125308/model-00003-1.48982-0.18571-1.59600-0.33000.keras

Epoch 4: saving model to model_init_2025-01-0305_13_55.125308/model-00004-1.48447-0.23000-1.57280-0.23000.keras

Epoch 5: saving model to model_init_2025-01-0305_13_55.125308/model-00005-1.46805-0.24000-1.56003-0.23000.keras

Epoch 6: saving model to model_init_2025-01-0305_13_55.125308/model-00006-1.45040-0.22857-1.51494-0.28000.keras

Epoch 7: saving model to model_init_2025-01-0305_13_55.125308/model-00007-1.40940-0.28286-1.44939-0.31000.keras

Epoch 8: saving model to model_init_2025-01-0305_13_55.125308/model-00008-1.35961-0.31571-1.35452-0.36000.keras

Epoch 9: saving model to model_init_2025-01-0305_13_55.125308/model-00009-1.29978-0.33000-1.274

### for this model train accuracy seem to be increasing with every epoch , but validation accuarcy seem to be unstable ,

### **Model 3**

In [ ]:
img_idx = [i for i in range(0,30,2)]
# img_idx = [i for i in range(0,30,1)]
len(img_idx)

15

In [ ]:
model = Sequential()

model.add(Conv3D(32, (3,3,3), padding='same', activation='relu', strides=(1, 2, 2), input_shape=(len(img_idx),120, 160, 3)))
model.add(MaxPooling3D(pool_size=(2, 2, 2)))
model.add(Dropout(0.25))

model.add(Conv3D(64, (3,3,3), padding='same', activation='relu', strides=(1, 2, 2)))
model.add(MaxPooling3D(pool_size=(2, 2, 2)))
model.add(Dropout(0.25))

model.add(Conv3D(128, (3,3,3), padding='same', activation='relu', strides=(1, 2, 2)))
model.add(MaxPooling3D(pool_size=(2, 2, 2)))
model.add(Dropout(0.25))

# model.add(Conv3D(256, (3,3,3), padding='same', activation='relu'))
# model.add(MaxPooling3D(pool_size=(2, 2, 2)))
# model.add(Dropout(0.25))

# model.add(Flatten())
model.add(GlobalAveragePooling3D())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(5, activation='softmax'))

In [ ]:
fitmodel(model)

Model: "sequential_17"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv3d_39 (Conv3D)          (None, 15, 60, 80, 32)    2624      
                                                                 
 max_pooling3d_39 (MaxPooli  (None, 7, 30, 40, 32)     0         
 ng3D)                                                           
                                                                 
 dropout_54 (Dropout)        (None, 7, 30, 40, 32)     0         
                                                                 
 conv3d_40 (Conv3D)          (None, 7, 15, 20, 64)     55360     
                                                                 
 max_pooling3d_40 (MaxPooli  (None, 3, 7, 10, 64)      0         
 ng3D)                                                           
                                                                 
 dropout_55 (Dropout)        (None, 3, 7, 10, 64)    

<ipython-input-69-b5c9ef93ece2>:29: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)
<ipython-input-69-b5c9ef93ece2>:54: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)



Epoch 1: saving model to model_init_2025-01-0305_13_55.125308/model-00001-1.49916-0.18143-1.58033-0.23000.keras

Epoch 2: saving model to model_init_2025-01-0305_13_55.125308/model-00002-1.48011-0.18571-1.52212-0.29000.keras

Epoch 3: saving model to model_init_2025-01-0305_13_55.125308/model-00003-1.40639-0.25857-1.46762-0.45000.keras

Epoch 4: saving model to model_init_2025-01-0305_13_55.125308/model-00004-1.34377-0.36571-1.33039-0.46000.keras

Epoch 5: saving model to model_init_2025-01-0305_13_55.125308/model-00005-1.24534-0.39429-0.55912-0.25000.keras

Epoch 6: saving model to model_init_2025-01-0305_13_55.125308/model-00006-1.12654-0.43857-0.58569-0.21000.keras

Epoch 7: saving model to model_init_2025-01-0305_13_55.125308/model-00007-1.08862-0.47857-0.49533-0.30000.keras

Epoch 8: saving model to model_init_2025-01-0305_13_55.125308/model-00008-0.94518-0.52286-0.37892-0.31000.keras

Epoch 9: saving model to model_init_2025-01-0305_13_55.125308/model-00009-0.84771-0.57857-0.290

### Validation accuracy still unstable,but traning accuracy has improved considerably

Trying with bigger batch size

In [ ]:
train_generator = generator(train_path, train_doc, 100)
val_generator = generator(val_path, val_doc, 100)

In [ ]:
fitmodel(model)

Model: "sequential_18"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv3d_42 (Conv3D)          (None, 15, 60, 80, 32)    2624      
                                                                 
 max_pooling3d_42 (MaxPooli  (None, 7, 30, 40, 32)     0         
 ng3D)                                                           
                                                                 
 dropout_58 (Dropout)        (None, 7, 30, 40, 32)     0         
                                                                 
 conv3d_43 (Conv3D)          (None, 7, 15, 20, 64)     55360     
                                                                 
 max_pooling3d_43 (MaxPooli  (None, 3, 7, 10, 64)      0         
 ng3D)                                                           
                                                                 
 dropout_59 (Dropout)        (None, 3, 7, 10, 64)    

<ipython-input-69-b5c9ef93ece2>:29: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)
<ipython-input-69-b5c9ef93ece2>:54: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)



Epoch 1: saving model to model_init_2025-01-0305_13_55.125308/model-00001-1.50548-0.21000-1.59664-0.23000.keras

Epoch 2: saving model to model_init_2025-01-0305_13_55.125308/model-00002-1.48817-0.18429-1.57141-0.23000.keras

Epoch 3: saving model to model_init_2025-01-0305_13_55.125308/model-00003-1.44650-0.18143-1.46805-0.29000.keras

Epoch 4: saving model to model_init_2025-01-0305_13_55.125308/model-00004-1.38422-0.28857-0.72433-0.21000.keras

Epoch 5: saving model to model_init_2025-01-0305_13_55.125308/model-00005-1.28049-0.41714-0.54752-0.32000.keras

Epoch 6: saving model to model_init_2025-01-0305_13_55.125308/model-00006-1.16049-0.44143-0.61000-0.20000.keras

Epoch 7: saving model to model_init_2025-01-0305_13_55.125308/model-00007-1.01668-0.50000-0.48983-0.29000.keras

Epoch 8: saving model to model_init_2025-01-0305_13_55.125308/model-00008-0.96229-0.52286-0.39776-0.30000.keras

Epoch 9: saving model to model_init_2025-01-0305_13_55.125308/model-00009-0.81257-0.60429-0.627

Trying with smaller batch size

In [ ]:
train_generator = generator(train_path, train_doc, 10)
val_generator = generator(val_path, val_doc, 10)

In [ ]:
fitmodel(model)

Model: "sequential_18"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv3d_42 (Conv3D)          (None, 15, 60, 80, 32)    2624      
                                                                 
 max_pooling3d_42 (MaxPooli  (None, 7, 30, 40, 32)     0         
 ng3D)                                                           
                                                                 
 dropout_58 (Dropout)        (None, 7, 30, 40, 32)     0         
                                                                 
 conv3d_43 (Conv3D)          (None, 7, 15, 20, 64)     55360     
                                                                 
 max_pooling3d_43 (MaxPooli  (None, 3, 7, 10, 64)      0         
 ng3D)                                                           
                                                                 
 dropout_59 (Dropout)        (None, 3, 7, 10, 64)    

<ipython-input-69-b5c9ef93ece2>:29: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)
<ipython-input-69-b5c9ef93ece2>:54: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)



Epoch 1: saving model to model_init_2025-01-0305_13_55.125308/model-00001-0.56404-0.74714-0.45100-0.79000.keras

Epoch 2: saving model to model_init_2025-01-0305_13_55.125308/model-00002-0.41099-0.78429-0.42943-0.84000.keras

Epoch 3: saving model to model_init_2025-01-0305_13_55.125308/model-00003-0.38775-0.79714-0.38373-0.84000.keras

Epoch 4: saving model to model_init_2025-01-0305_13_55.125308/model-00004-0.47039-0.78000-0.40788-0.84000.keras

Epoch 5: saving model to model_init_2025-01-0305_13_55.125308/model-00005-0.40707-0.79857-0.40356-0.84000.keras

Epoch 6: saving model to model_init_2025-01-0305_13_55.125308/model-00006-0.37704-0.80714-0.36470-0.87000.keras

Epoch 7: saving model to model_init_2025-01-0305_13_55.125308/model-00007-0.44315-0.79143-0.37410-0.34000.keras

Epoch 8: saving model to model_init_2025-01-0305_13_55.125308/model-00008-0.66735-0.75571-0.20225-0.45000.keras

Epoch 9: saving model to model_init_2025-01-0305_13_55.125308/model-00009-0.58204-0.76714-0.185

### **Model 4**

### Trying with different stride size

In [33]:
# img_idx = [i for i in range(0,30,2)]
img_idx = [i for i in range(0,30,1)]
len(img_idx)

30

In [34]:
model = Sequential()

model.add(Conv3D(32, (3,3,3), padding='same', activation='relu', strides=(1, 2, 2), input_shape=(len(img_idx),120, 160, 3)))
model.add(MaxPooling3D(pool_size=(2, 2, 2)))
model.add(Dropout(0.25))

model.add(Conv3D(64, (3,3,3), padding='same', activation='relu', strides=(1, 2, 2)))
model.add(MaxPooling3D(pool_size=(2, 2, 2)))
model.add(Dropout(0.25))

model.add(Conv3D(128, (3,3,3), padding='same', activation='relu', strides=(1, 2, 2)))
model.add(MaxPooling3D(pool_size=(2, 2, 2)))
model.add(Dropout(0.25))

# model.add(Conv3D(256, (3,3,3), padding='same', activation='relu'))
# model.add(MaxPooling3D(pool_size=(2, 2, 2)))
# model.add(Dropout(0.25))

# model.add(Flatten())
model.add(GlobalAveragePooling3D())
model.add(Dense(256, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(5, activation='softmax'))

In [35]:
fitmodel(model)

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv3d (Conv3D)             (None, 30, 60, 80, 32)    2624      
                                                                 
 max_pooling3d (MaxPooling3  (None, 15, 30, 40, 32)    0         
 D)                                                              
                                                                 
 dropout_3 (Dropout)         (None, 15, 30, 40, 32)    0         
                                                                 
 conv3d_1 (Conv3D)           (None, 15, 15, 20, 64)    55360     
                                                                 
 max_pooling3d_1 (MaxPoolin  (None, 7, 7, 10, 64)      0         
 g3D)                                                            
                                                                 
 dropout_4 (Dropout)         (None, 7, 7, 10, 64)      0

<ipython-input-32-0ab3b20f755c>:29: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)


KeyboardInterrupt: 

### Validation accuracy still unstable,but traning accuracy has also worserned

### **Model5**

In [22]:
model = Sequential()

model.add(Conv3D(64, (3,3,3), padding='same', activation='relu', strides=(1, 3, 3), input_shape=(len(img_idx),120, 160, 3)))
model.add(MaxPooling3D(pool_size=(2, 2, 2)))
model.add(Dropout(0.25))

model.add(Conv3D(128, (3,3,3), padding='same', activation='relu', strides=(1, 3, 3)))
model.add(MaxPooling3D(pool_size=(2, 2, 2)))
model.add(Dropout(0.25))

# model.add(Conv3D(128, (3,3,3), padding='same', activation='relu', strides=(1, 2, 2)))
# model.add(MaxPooling3D(pool_size=(2, 2, 2)))
# model.add(Dropout(0.25))

# model.add(Conv3D(256, (3,3,3), padding='same', activation='relu'))
# model.add(MaxPooling3D(pool_size=(2, 2, 2)))
# model.add(Dropout(0.25))

# model.add(Flatten())
model.add(GlobalAveragePooling3D())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(5, activation='softmax'))

In [24]:
fitmodel(model)

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv3d_6 (Conv3D)           (None, 15, 40, 54, 64)    5248      
                                                                 
 max_pooling3d_6 (MaxPoolin  (None, 7, 20, 27, 64)     0         
 g3D)                                                            
                                                                 
 dropout_9 (Dropout)         (None, 7, 20, 27, 64)     0         
                                                                 
 conv3d_7 (Conv3D)           (None, 7, 7, 9, 128)      221312    
                                                                 
 max_pooling3d_7 (MaxPoolin  (None, 3, 3, 4, 128)      0         
 g3D)                                                            
                                                                 
 dropout_10 (Dropout)        (None, 3, 3, 4, 128)     

<ipython-input-8-b5c9ef93ece2>:29: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)
<ipython-input-8-b5c9ef93ece2>:54: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)



Epoch 1: saving model to model_init_2025-01-0406_10_04.796943/model-00001-1.50234-0.19143-1.58801-0.23000.keras

Epoch 2: saving model to model_init_2025-01-0406_10_04.796943/model-00002-1.48718-0.19571-1.57331-0.24000.keras

Epoch 3: saving model to model_init_2025-01-0406_10_04.796943/model-00003-1.46463-0.19429-1.52616-0.28000.keras

Epoch 4: saving model to model_init_2025-01-0406_10_04.796943/model-00004-1.42423-0.23857-1.46521-0.28000.keras

Epoch 5: saving model to model_init_2025-01-0406_10_04.796943/model-00005-1.37596-0.29571-1.48093-0.25000.keras

Epoch 6: saving model to model_init_2025-01-0406_10_04.796943/model-00006-1.36136-0.29571-1.34053-0.46000.keras

Epoch 7: saving model to model_init_2025-01-0406_10_04.796943/model-00007-1.27257-0.35429-1.26401-0.54000.keras

Epoch 8: saving model to model_init_2025-01-0406_10_04.796943/model-00008-1.20979-0.39286-1.27274-0.50000.keras

Epoch 9: saving model to model_init_2025-01-0406_10_04.796943/model-00009-1.16331-0.42571-1.121

###Transfer Learning and LSTM

In [11]:
img_idx = [i for i in range(0,30,2)]
# img_idx = [i for i in range(0,30,1)]
len(img_idx)

15

In [24]:
num_epochs = 20

In [36]:
from tensorflow.keras.applications import ResNet50,MobileNetV2
from tensorflow.keras.layers import Input, TimeDistributed, GRU, Dense,GlobalAveragePooling2D
from tensorflow.keras.models import Model

input_layer = Input(shape=(len(img_idx), 120, 160, 3))

# base_model = ResNet50(weights='imagenet', include_top=False, input_shape=(120, 160, 3))
base_model = MobileNetV2(weights='imagenet', include_top=False, input_shape=(120, 160, 3)) # or MobileNet

for layer in base_model.layers:
    layer.trainable = False

x = TimeDistributed(base_model)(input_layer)
x = TimeDistributed(GlobalAveragePooling2D())(x)

x = GRU(128)(x)

output_layer = Dense(5, activation='softmax')(x)

model = Model(inputs=input_layer, outputs=output_layer)


<ipython-input-32-0ab3b20f755c>:29: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)


In [37]:
fitmodel(model)

Model: "model_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_7 (InputLayer)        [(None, 30, 120, 160, 3   0         
                             )]                                  
                                                                 
 time_distributed_6 (TimeDi  (None, 30, 4, 5, 1280)    2257984   
 stributed)                                                      
                                                                 
 time_distributed_7 (TimeDi  (None, 30, 1280)          0         
 stributed)                                                      
                                                                 
 gru_4 (GRU)                 (None, 128)               541440    
                                                                 
 dense_5 (Dense)             (None, 5)                 645       
                                                           

<ipython-input-32-0ab3b20f755c>:29: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)
<ipython-input-32-0ab3b20f755c>:53: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)



Epoch 1: saving model to model_init_2025-01-0719_31_49.888542/model-00001-1.28318-0.41571-1.01884-0.57000.keras

Epoch 2: saving model to model_init_2025-01-0719_31_49.888542/model-00002-0.56465-0.74000-0.32527-0.41000.keras

Epoch 3: saving model to model_init_2025-01-0719_31_49.888542/model-00003-0.32214-0.84429-0.33627-0.35000.keras

Epoch 4: saving model to model_init_2025-01-0719_31_49.888542/model-00004-0.20018-0.89286-0.26356-0.39000.keras

Epoch 5: saving model to model_init_2025-01-0719_31_49.888542/model-00005-0.12811-0.91429-0.24796-0.41000.keras

Epoch 6: saving model to model_init_2025-01-0719_31_49.888542/model-00006-0.09871-0.92429-0.23551-0.43000.keras

Epoch 7: saving model to model_init_2025-01-0719_31_49.888542/model-00007-0.07419-0.92714-0.23048-0.40000.keras

Epoch 8: saving model to model_init_2025-01-0719_31_49.888542/model-00008-0.04991-0.92857-0.26636-0.37000.keras

Epoch 9: saving model to model_init_2025-01-0719_31_49.888542/model-00009-0.04415-0.92857-0.272

### Validation accuracy seem to be unstable

###Adding one more GSU layer

In [38]:
input_layer = Input(shape=(len(img_idx), 120, 160, 3))

# base_model = ResNet50(weights='imagenet', include_top=False, input_shape=(120, 160, 3))
base_model = MobileNetV2(weights='imagenet', include_top=False, input_shape=(120, 160, 3)) # or MobileNet

for layer in base_model.layers:
    layer.trainable = False

x = TimeDistributed(base_model)(input_layer)
x = TimeDistributed(GlobalAveragePooling2D())(x)

x = GRU(128, return_sequences=True)(x)
x = Dropout(0.25)(x) # Added dropout for regularization

# Second GRU layer to further process the sequential information
x = GRU(64)(x)
x = Dropout(0.25)(x) # Added dropout for regularization

output_layer = Dense(5, activation='softmax')(x)

model = Model(inputs=input_layer, outputs=output_layer)

In [41]:
fitmodel(model)

Model: "model_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_7 (InputLayer)        [(None, 30, 120, 160, 3   0         
                             )]                                  
                                                                 
 time_distributed_6 (TimeDi  (None, 30, 4, 5, 1280)    2257984   
 stributed)                                                      
                                                                 
 time_distributed_7 (TimeDi  (None, 30, 1280)          0         
 stributed)                                                      
                                                                 
 gru_4 (GRU)                 (None, 30, 128)           541440    
                                                                 
 dropout_12 (Dropout)        (None, 30, 128)           0         
                                                           

<ipython-input-40-c9a905daff97>:29: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)
<ipython-input-40-c9a905daff97>:53: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)



Epoch 1: saving model to model_init_2025-01-0406_10_04.796943/model-00001-1.27043-0.47892-0.99485-0.57000.keras

Epoch 2: saving model to model_init_2025-01-0406_10_04.796943/model-00002-0.73750-0.70633-0.58880-0.69000.keras

Epoch 3: saving model to model_init_2025-01-0406_10_04.796943/model-00003-0.56575-0.76506-0.60816-0.68000.keras

Epoch 4: saving model to model_init_2025-01-0406_10_04.796943/model-00004-0.38464-0.85542-0.53727-0.70000.keras

Epoch 5: saving model to model_init_2025-01-0406_10_04.796943/model-00005-0.34601-0.87801-0.47299-0.81000.keras

Epoch 6: saving model to model_init_2025-01-0406_10_04.796943/model-00006-0.29938-0.88705-0.74736-0.70000.keras

Epoch 7: saving model to model_init_2025-01-0406_10_04.796943/model-00007-0.23184-0.91717-0.48198-0.80000.keras

Epoch 8: saving model to model_init_2025-01-0406_10_04.796943/model-00008-0.17924-0.93223-0.67721-0.75000.keras

Epoch 8: ReduceLROnPlateau reducing learning rate to 0.0005000000237487257.

Epoch 9: saving mo

### above model seems to be overfitting

###Reverting back to models with single gsu layer to simplfy due to overfitting

### Regularization as overfitting

In [30]:
input_layer = Input(shape=(len(img_idx), 120, 160, 3))

# base_model = ResNet50(weights='imagenet', include_top=False, input_shape=(120, 160, 3))
base_model = MobileNetV2(weights='imagenet', include_top=False, input_shape=(120, 160, 3)) # or MobileNet

for layer in base_model.layers:
    layer.trainable = True

x = TimeDistributed(base_model)(input_layer)
x = TimeDistributed(GlobalAveragePooling2D())(x)

x = GRU(128)(x)
x = BatchNormalization()(x)
x = Dropout(0.50)(x) # Added dropout for regularization

output_layer = Dense(5, activation='softmax')(x)

model = Model(inputs=input_layer, outputs=output_layer)

In [28]:
fitmodel(model)

Model: "model_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_15 (InputLayer)       [(None, 15, 120, 160, 3   0         
                             )]                                  
                                                                 
 time_distributed_14 (TimeD  (None, 15, 4, 5, 1280)    2257984   
 istributed)                                                     
                                                                 
 time_distributed_15 (TimeD  (None, 15, 1280)          0         
 istributed)                                                     
                                                                 
 gru_7 (GRU)                 (None, 128)               541440    
                                                                 
 batch_normalization_5 (Bat  (None, 128)               512       
 chNormalization)                                          

<ipython-input-9-c9a905daff97>:29: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)
<ipython-input-9-c9a905daff97>:53: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)



Epoch 1: saving model to model_init_2025-01-0805_08_00.479798/model-00001-1.02778-0.60857-1.88065-0.39000.keras

Epoch 2: saving model to model_init_2025-01-0805_08_00.479798/model-00002-0.24425-0.89286-1.94187-0.12000.keras

Epoch 3: saving model to model_init_2025-01-0805_08_00.479798/model-00003-0.10627-0.94857-1.32267-0.21000.keras

Epoch 4: saving model to model_init_2025-01-0805_08_00.479798/model-00004-0.10304-0.97857-1.27452-0.20000.keras

Epoch 5: saving model to model_init_2025-01-0805_08_00.479798/model-00005-0.03994-0.94857-1.04761-0.17000.keras

Epoch 6: saving model to model_init_2025-01-0805_08_00.479798/model-00006-0.04257-0.92857-0.95666-0.22000.keras

Epoch 7: saving model to model_init_2025-01-0805_08_00.479798/model-00007-0.02440-0.93429-0.84150-0.20000.keras

Epoch 8: saving model to model_init_2025-01-0805_08_00.479798/model-00008-0.03288-0.94714-0.68966-0.27000.keras

Epoch 9: saving model to model_init_2025-01-0805_08_00.479798/model-00009-0.00824-0.94000-0.798

### Model overfitting a lot ,Trying with diff batch size

In [31]:
batch_size = 10

50

In [36]:
input_layer = Input(shape=(len(img_idx), 120, 160, 3))

# base_model = ResNet50(weights='imagenet', include_top=False, input_shape=(120, 160, 3))
base_model = MobileNetV2(weights='imagenet', include_top=False, input_shape=(120, 160, 3)) # or MobileNet

for layer in base_model.layers:
    layer.trainable = False

x = TimeDistributed(base_model)(input_layer)
x = TimeDistributed(GlobalAveragePooling2D())(x)

x = GRU(128, kernel_regularizer=regularizers.l2(0.01))(x)
# x = BatchNormalization()(x)
x = Dropout(0.50)(x) # Added dropout for regularization

output_layer = Dense(5, activation='softmax')(x)

model = Model(inputs=input_layer, outputs=output_layer)

In [37]:
fitmodel(model)

Model: "model_6"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_25 (InputLayer)       [(None, 15, 120, 160, 3   0         
                             )]                                  
                                                                 
 time_distributed_23 (TimeD  (None, 15, 4, 5, 1280)    2257984   
 istributed)                                                     
                                                                 
 time_distributed_24 (TimeD  (None, 15, 1280)          0         
 istributed)                                                     
                                                                 
 gru_10 (GRU)                (None, 128)               541440    
                                                                 
 dropout_4 (Dropout)         (None, 128)               0         
                                                           

<ipython-input-9-c9a905daff97>:29: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)
<ipython-input-9-c9a905daff97>:53: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)



Epoch 1: saving model to model_init_2025-01-0807_56_27.727542/model-00001-6.71957-0.34000-5.28926-0.68000.keras

Epoch 2: saving model to model_init_2025-01-0807_56_27.727542/model-00002-4.68285-0.59857-3.47952-0.42000.keras

Epoch 3: saving model to model_init_2025-01-0807_56_27.727542/model-00003-3.43592-0.71714-2.78628-0.38000.keras

Epoch 4: saving model to model_init_2025-01-0807_56_27.727542/model-00004-2.68069-0.80714-2.24164-0.38000.keras

Epoch 5: saving model to model_init_2025-01-0807_56_27.727542/model-00005-2.19141-0.83571-1.96815-0.39000.keras

Epoch 6: saving model to model_init_2025-01-0807_56_27.727542/model-00006-1.84079-0.87857-1.69442-0.44000.keras

Epoch 7: saving model to model_init_2025-01-0807_56_27.727542/model-00007-1.58704-0.89143-1.57031-0.38000.keras

Epoch 8: saving model to model_init_2025-01-0807_56_27.727542/model-00008-1.38157-0.90714-1.42619-0.39000.keras

Epoch 9: saving model to model_init_2025-01-0807_56_27.727542/model-00009-1.24069-0.90143-1.307

### No improvement with batch size,Trying less complicated model using jus 64 neurons in GRU

In [38]:
input_layer = Input(shape=(len(img_idx), 120, 160, 3))

# base_model = ResNet50(weights='imagenet', include_top=False, input_shape=(120, 160, 3))
base_model = MobileNetV2(weights='imagenet', include_top=False, input_shape=(120, 160, 3)) # or MobileNet

for layer in base_model.layers:
    layer.trainable = False

x = TimeDistributed(base_model)(input_layer)
x = TimeDistributed(GlobalAveragePooling2D())(x)

x = GRU(64, kernel_regularizer=regularizers.l2(0.01))(x)
# x = BatchNormalization()(x)
x = Dropout(0.50)(x) # Added dropout for regularization

output_layer = Dense(5, activation='softmax')(x)

model = Model(inputs=input_layer, outputs=output_layer)

In [39]:
fitmodel(model)

Model: "model_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_27 (InputLayer)       [(None, 15, 120, 160, 3   0         
                             )]                                  
                                                                 
 time_distributed_25 (TimeD  (None, 15, 4, 5, 1280)    2257984   
 istributed)                                                     
                                                                 
 time_distributed_26 (TimeD  (None, 15, 1280)          0         
 istributed)                                                     
                                                                 
 gru_11 (GRU)                (None, 64)                258432    
                                                                 
 dropout_5 (Dropout)         (None, 64)                0         
                                                           

<ipython-input-9-c9a905daff97>:29: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)
<ipython-input-9-c9a905daff97>:53: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)



Epoch 1: saving model to model_init_2025-01-0808_49_04.597781/model-00001-4.61864-0.33571-3.72226-0.56000.keras

Epoch 2: saving model to model_init_2025-01-0808_49_04.597781/model-00002-3.25903-0.59286-2.44323-0.36000.keras

Epoch 3: saving model to model_init_2025-01-0808_49_04.597781/model-00003-2.53707-0.68571-2.05585-0.84000.keras

Epoch 4: saving model to model_init_2025-01-0808_49_04.597781/model-00004-2.04310-0.75714-1.68637-0.35000.keras

Epoch 5: saving model to model_init_2025-01-0808_49_04.597781/model-00005-1.70672-0.80571-1.50412-0.87000.keras

Epoch 6: saving model to model_init_2025-01-0808_49_04.597781/model-00006-1.44745-0.85429-1.40465-0.87000.keras

Epoch 7: saving model to model_init_2025-01-0808_49_04.597781/model-00007-1.31635-0.84286-1.24001-0.38000.keras

Epoch 8: saving model to model_init_2025-01-0808_49_04.597781/model-00008-1.17479-0.87143-1.10767-0.89000.keras

Epoch 9: saving model to model_init_2025-01-0808_49_04.597781/model-00009-1.04157-0.89000-0.988

### Validation accuracy unsable , model overfitting for many epochs,Trying lesser complicated model using jus 32 neurons in GRU

In [41]:
input_layer = Input(shape=(len(img_idx), 120, 160, 3))

# base_model = ResNet50(weights='imagenet', include_top=False, input_shape=(120, 160, 3))
base_model = MobileNetV2(weights='imagenet', include_top=False, input_shape=(120, 160, 3)) # or MobileNet

for layer in base_model.layers:
    layer.trainable = False

x = TimeDistributed(base_model)(input_layer)
x = TimeDistributed(GlobalAveragePooling2D())(x)

x = GRU(32, kernel_regularizer=regularizers.l2(0.01))(x)
# x = BatchNormalization()(x)
x = Dropout(0.50)(x) # Added dropout for regularization

output_layer = Dense(5, activation='softmax')(x)

model = Model(inputs=input_layer, outputs=output_layer)

In [43]:
fitmodel(model)

Model: "model_8"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_29 (InputLayer)       [(None, 15, 120, 160, 3   0         
                             )]                                  
                                                                 
 time_distributed_27 (TimeD  (None, 15, 4, 5, 1280)    2257984   
 istributed)                                                     
                                                                 
 time_distributed_28 (TimeD  (None, 15, 1280)          0         
 istributed)                                                     
                                                                 
 gru_12 (GRU)                (None, 32)                126144    
                                                                 
 dropout_6 (Dropout)         (None, 32)                0         
                                                           

<ipython-input-9-c9a905daff97>:29: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)
<ipython-input-9-c9a905daff97>:53: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)



Epoch 1: saving model to model_init_2025-01-0812_22_57.545227/model-00001-0.59694-0.85571-0.88421-0.80000.keras

Epoch 2: saving model to model_init_2025-01-0812_22_57.545227/model-00002-0.47123-0.90714-0.54305-0.91000.keras

Epoch 3: saving model to model_init_2025-01-0812_22_57.545227/model-00003-0.41759-0.92857-0.49905-0.93000.keras

Epoch 4: saving model to model_init_2025-01-0812_22_57.545227/model-00004-0.40984-0.93000-0.58636-0.39000.keras

Epoch 5: saving model to model_init_2025-01-0812_22_57.545227/model-00005-0.38971-0.93286-0.56884-0.85000.keras

Epoch 6: saving model to model_init_2025-01-0812_22_57.545227/model-00006-0.35083-0.93000-0.53126-0.41000.keras

Epoch 6: ReduceLROnPlateau reducing learning rate to 0.0005000000237487257.

Epoch 7: saving model to model_init_2025-01-0812_22_57.545227/model-00007-0.33785-0.94429-0.46204-0.42000.keras

Epoch 8: saving model to model_init_2025-01-0812_22_57.545227/model-00008-0.31581-0.94571-0.51552-0.92000.keras

Epoch 9: saving mo

### Validation accuracy seem stable for most of the epochs, diffrence betten train and val accuracy seem to be very less, so no over fitting, thus considering this as our best model

### Evaluation on best model

In [51]:
import numpy as np
from tensorflow.keras.models import load_model
model = load_model('/content/model_init_2025-01-0812_22_57.545227/model-00020-0.23830-0.94714-0.39164-0.93000.keras')



In [52]:
val_generator = generator(val_path, val_doc, batch_size=num_val_sequences)

loss, accuracy = model.evaluate(val_generator, steps=1)


print(f"Validation Loss: {loss:.4f}")
print(f"Validation Accuracy: {accuracy:.4f}")


<ipython-input-9-c9a905daff97>:29: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)


1/1 [==============================] - 8s 8s/step - loss: 0.7412 - categorical_accuracy: 0.8100
Validation Loss: 0.7412
Validation Accuracy: 0.8100


In [53]:
### /content/model_init_2025-01-0808_49_04.597781/model-00020-0.45582-0.93143-0.61357-0.92000.keras

model = load_model('/content/model_init_2025-01-0808_49_04.597781/model-00020-0.45582-0.93143-0.61357-0.92000.keras')

val_generator = generator(val_path, val_doc, batch_size=num_val_sequences)

loss, accuracy = model.evaluate(val_generator, steps=1)


print(f"Validation Loss: {loss:.4f}")
print(f"Validation Accuracy: {accuracy:.4f}")


<ipython-input-9-c9a905daff97>:29: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  image = imread(source_path+'/'+ t[folder + (batch*batch_size)].strip().split(';')[0]+'/'+imgs[item]).astype(np.float32)


1/1 [==============================] - 7s 7s/step - loss: 0.8018 - categorical_accuracy: 0.8600
Validation Loss: 0.8018
Validation Accuracy: 0.8600


### Accuracy on whole validation set came out to be 0.86 which we think is quite accpetable thus submitting above model